src/Event.h

```diff
@@ -0,0 +1,131 @@
+#pragma once
+
+#include <algorithm>
+#include <atomic>
+#include <cstddef>
+#include <functional>
+#include <iostream>
+#include <list>
+#include <mutex>
+#include <vector>
+
+template<typename... Args>
+class EventHandler {
+public:
+    typedef std::function<void(Args...)> HandlerFunc;
+    typedef size_t HandlerId;
+
+    explicit EventHandler(const HandlerFunc& func)
+        : m_HandlerFunc(func)
+    {
+        m_Id = ++s_IdCounter;
+    }
+
+    EventHandler(const EventHandler& handler) :
+        m_Id(handler.m_Id), m_HandlerFunc(handler.m_HandlerFunc)
+    {}
+
+    EventHandler(EventHandler&& handler)
+        : m_Id(handler.m_Id), m_HandlerFunc(handler.m_HandlerFunc)
+    {}
+
+    EventHandler& operator=(EventHandler& handler)
+    {
+        m_HandlerFunc = handler.m_HandlerFunc;
+        m_Id = handler.m_Id;
+        return *this;
+    }
+
+    EventHandler& operator=(EventHandler&& handler)
+    {
+        m_HandlerFunc = handler.m_HandlerFunc;
+        m_Id = handler.m_Id;
+        return *this;
+    }
+
+    void operator()(Args... params)
+    {
+        if (m_HandlerFunc) {
+            m_HandlerFunc(params...);
+        }
+    }
+
+    bool operator==(const EventHandler<Args...> handler)
+    {
+        return m_Id == handler.id();
+    }
+
+    HandlerId id() const { return m_Id; }
+
+private:
+    HandlerId m_Id;
+    HandlerFunc m_HandlerFunc;
+    static inline std::atomic<HandlerId> s_IdCounter = 0;
+};
+
+class EventBase {
+protected:
+    EventBase(size_t type) : m_Type(type) {}
+    size_t m_Type;
+};
+
+//we will have subclass inherit on this
+template<typename ...Args>
+class Event : public EventBase {
+    typedef EventHandler<Args...> HandlerType;
+public:
+    Event(size_t type) : EventBase(type) {}
+
+    void emit(Args ... args)
```

Comment:
Upper case function names. 
```suggestion
    void Emit(Args ... args)
```